middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -620,20 +620,12 @@ let bytes_like_set ~dbg ~unsafe
     check_access ~dbg ~size_int ~access_size ~primitive:unsafe_set bytes
       ~index_kind index
 
-(* Array vector load/store *)
+(* Array bounds checks *)
 
-let array_vector_access_validity_condition array ~size_int
-    (array_kind : P.Array_kind.t) index =
-  let width_in_scalars =
-    match array_kind with
-    | Naked_floats | Immediates | Naked_int64s | Naked_nativeints -> 2
-    | Naked_int32s | Naked_float32s -> 4
-    | Values ->
-      Misc.fatal_error
-        "Attempted to load/store a SIMD vector from/to a value array."
-  in
+let multiple_word_array_access_validity_condition array ~size_int
+    array_length_kind index_kind ~width_in_scalars ~index =
   let length_untagged =
-    untag_int (H.Prim (Unary (Array_length (Array_kind array_kind), array)))
+    untag_int (H.Prim (Unary (Array_length array_length_kind, array)))
   in
   let reduced_length_untagged =
```

Comment:
We should special-case `width_in_scalars=1` to be sure normal array accesses won't generate unecessary math here